Функция `Формат`.

dmpas opened the pull request:
Closes: #125

dmpas committed:
Функция `Формат`.

dmpas committed:
Замечания по Сонару.

dmpas committed:
Замечания по Сонару.

dmpas committed:
Замечания по Сонару.

dmpas committed:
Замечания по Сонару.

dmpas committed:
Замечания по Сонару.

dmpas committed:
Покрытие тестами.

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java:
```diff
@@ -0,0 +1,145 @@
+  }
+
+  public String get(String[] names) {
+    for (String n : names) {
```
@dmpas давай дадим `n` более осмысленное имя

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java:
```diff
@@ -0,0 +1,145 @@
+        return paramList.get(n.toUpperCase());
+      }
+    }
+    return null;
```
@dmpas давай обойдемся без `null`. Потом с результатом работы метода легко ошибиться.

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java:
```diff
@@ -0,0 +1,145 @@
+    return null;
+  }
+
+  private int index;
```
@dmpas объявление филдов до конструктора

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java:
```diff
@@ -0,0 +1,145 @@
+    final var start = index;
+
+    while (index < format.length()) {
+      final var c = format.charAt(index);
```
@dmpas давай дадим `c` более осмысленное имя

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java:
```diff
@@ -0,0 +1,145 @@
+      return;
+    }
+
+    final var sb = new StringBuilder();
```
@dmpas давай дадим `sb` более осмысленное имя

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java:
```diff
@@ -0,0 +1,145 @@
+      index++;
+    }
+    while (index < format.length()) {
+      final var c = format.charAt(index);
```
@dmpas давай дадим c более осмысленное имя

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/NumberFormatter.java:
```diff
@@ -0,0 +1,219 @@
+
+public class NumberFormatter {
+
+  public static final int NEGATIVE_BRACES = 0; // (1,1)
```
@dmpas если используется только в рамках этого класса, нужно сделать private

otymko replied:
NEGATIVE_* константы очень похожи на enum.

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/NumberFormatter.java:
```diff
@@ -0,0 +1,219 @@
+  }
+
+  @Getter
+  private int maxSize = -1;
```
@dmpas филды раньше конструктора / методов

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/NumberFormatter.java:
```diff
@@ -0,0 +1,219 @@
+    }
+    final var group1pos = value.length() - lowerGroupingSize;
+
+    final var a = new ArrayList<String>();
```
@dmpas давай дадим `a` более осмысленное имя

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/NumberFormatter.java:
```diff
@@ -0,0 +1,219 @@
+      }
+    }
+
+    if (!rest.isEmpty()) a.add(rest);
```
@dmpas давай не будем использовать однострочные if

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/NumberFormatter.java:
```diff
@@ -0,0 +1,219 @@
+      return zeroAppearance;
+    }
+
+    final var systemFormat = (DecimalFormat)NumberFormat.getInstance(Locale.ROOT);
```
@dmpas Locale.ROOT ?

dmpas replied:
Локаль с предсказуемым поведением. Важно знать, что используются десятичные цифры и точка как разделитель.

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/NumberFormatter.java:
```diff
@@ -0,0 +1,219 @@
+      systemFormat.setMinimumFractionDigits(decimalSize);
+      systemFormat.setRoundingMode(RoundingMode.HALF_UP);
+    } else {
+      systemFormat.setMaximumFractionDigits(340);
```
@dmpas 340 в константы

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+  }
+
+  private static String boolFormat(boolean value, FormatParametersList params) {
+    final var p = params.get( value ? BOOLEAN_TRUE : BOOLEAN_FALSE);
```
@dmpas давай дадим `p` более осмысленное имя

otymko replied:
Как раз здесь бы хорошо смотрелся Optional из params.get()

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+      default:
+        throw MachineException.operationNotSupportedException();
+    }
+    if (formattedValue == null) return value.asString();
```
@dmpas давай не будем использовать однострочные if

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+
+  private static String convertToNativeFormat(String param) {
+    final var builder = new StringBuilder(param);
+    final var nativeMap = new HashMap<Character, Character>();
```
@dmpas давай вынесем в статик финал поле класса

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+
+  private static int parseInt(String param) {
+    final var sb = new StringBuilder();
+    for (var c : param.toCharArray()) {
```
@dmpas давай дадим `c` более осмысленное имя

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+  }
+
+  private static int parseInt(String param) {
+    final var sb = new StringBuilder();
```
@dmpas давай дадим `sb` более осмысленное имя

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+
+  private static String numberFormat(BigDecimal value, FormatParametersList params) {
+
+    final var nf = new NumberFormatter(getLocale(params.get(LOCALE)));
```
@dmpas давай дадим `nf` более осмысленное имя

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+
+    final var nf = new NumberFormatter(getLocale(params.get(LOCALE)));
+
+    final var nz = params.get(NUM_ZERO_APPEARANCE);
```
@dmpas давай дадим `nz` более осмысленное имя

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+      nf.setZeroAppearance(nz.equals("") ? "0" : nz);
+    }
+
+    final var nfd = params.get(NUM_DECIMAL_SIZE);
```
@dmpas давай дадим `nfd` более осмысленное имя

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+    final var nfd = params.get(NUM_DECIMAL_SIZE);
+    if (nfd != null && !nfd.isBlank()) {
+      final var i_nfd = parseInt(nfd);
+      if (i_nfd >= 0)
```
@dmpas { }

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+
+    nf.setLeadingZeroes(params.get(NUM_LEADING_ZERO) != null);
+
+    final var nd = params.get(NUM_MAX_SIZE);
```
@dmpas давай дадим `nd` более осмысленное имя

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+
+    final var nd = params.get(NUM_MAX_SIZE);
+    if (nd != null) {
+      final var i_nd = nd.isBlank() ? 0 : parseInt(nd);
```
@dmpas невалидное именование переменной `i_nd `

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+        nf.setMaxSize(i_nd);
+    }
+
+    final var nds = params.get(NUM_FRACTION_DELIMITER);
```
@dmpas давай дадим `nds` более осмысленное имя

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+      nf.setFractionDelimiter(nds.length() < 2 ? nds : nds.substring(0, 1));
+    }
+
+    final var ng = params.get(NUM_GROUPING);
```
@dmpas давай дадим `ng` более осмысленное имя

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+
+    final var ng = params.get(NUM_GROUPING);
+    if (ng != null) {
+      if (ng.equals("")) nf.setGroupingSize(0);
```
@dmpas давай не будет использовать однострочные `if`

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+      else {
+        final var groups = new ArrayList<Integer>();
+
+        for (var s : ng.split("\\D")) {
```
@dmpas давай дадим `s` более осмысленное имя

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+        final var groups = new ArrayList<Integer>();
+
+        for (var s : ng.split("\\D")) {
+          if (s.isBlank()) continue;
```
@dmpas давай не будем использовать однострочные if

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+          if (s.isBlank()) continue;
+          final var gv = parseInt(s);
+          groups.add(gv);
+          if (gv == 0) break;
```
@dmpas давай не будем использовать однострочные if

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+
+        for (var s : ng.split("\\D")) {
+          if (s.isBlank()) continue;
+          final var gv = parseInt(s);
```
@dmpas давай дадим `gv` более осмысленное имя

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+      }
+    }
+
+    final var nn = params.get(NUM_NEGATIVE_APPEARANCE);
```
@dmpas давай дадим `nn` более осмысленное имя

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+
+    final var nn = params.get(NUM_NEGATIVE_APPEARANCE);
+    if (nn != null) {
+      final var i_nn = parseInt(nn);
```
@dmpas невалидное именование переменной

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+      nf.setNegativeAppearance(i_nn);
+    }
+
+    final var ns = params.get(NUM_DECIMAL_SHIFT);
```
@dmpas давай дадим `ns` более осмысленное имя

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+
+    final var ns = params.get(NUM_DECIMAL_SHIFT);
+    if (ns != null) {
+      final var i_ns = parseInt(ns);
```
@dmpas невалидное именование переменной

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+      nf.setDecimalShift(i_ns);
+    }
+
+    final var ngs = params.get(NUM_GROUPS_DELIMITER);
```
@dmpas давай дадим `ngs` более осмысленное имя

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+    return nf.format(value);
+  }
+
+  private ValueFormatter() {}
```
@dmpas конструктор после филдов, но до методов

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java:
```diff
@@ -0,0 +1,145 @@
+  }
+
+  private static char getTerminalChar(char c) {
+    if (c == DOUBLE_QUOTE) return DOUBLE_QUOTE;
```
@dmpas давай не будем использовать однострочные if

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/NumberFormatter.java:
```diff
@@ -0,0 +1,219 @@
+
+  public void setMaxSize(int value) {
+    maxSize = value;
+    if (decimalSize == -1) decimalSize = 0;
```
@dmpas давай не будем использовать однострочные if

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/NumberFormatter.java:
```diff
@@ -0,0 +1,219 @@
+  }
+
+  private BigDecimal valueToFormat(BigDecimal value) {
+    if (decimalShift < 0) return value.movePointRight(-decimalShift);
```
@dmpas давай не будем использовать однострочные if

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,276 @@
+    return Integer.parseInt(sb.toString());
+  }
+
+  private static String numberFormat(BigDecimal value, FormatParametersList params) {
```
@dmpas получился очень тяжелый метод, давай его декомпозируем

dmpas committed:
Исправлено строковое представление даты.

dmpas committed:
Рефакторинг во имя Сонара.

dmpas committed:
Тест теперь локалезависимый.

dmpas committed:
Костыль для совместимости.

Краткое представление даты в en-US локали приведено в соответствие с поведением Вендора.

dmpas committed:
Попытка уменьшить когнитивную сложность.

dmpas committed:
Попытка уменьшить когнитивную сложность.

dmpas committed:
Приоритет форматов Дат.

theshadowco commented on engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java:
```diff
@@ -0,0 +1,216 @@
+import java.util.Map;
+import java.util.Optional;
+
+public class FormatParametersList {
```
где javadoc

theshadowco commented on engine/src/main/java/com/github/otymko/jos/runtime/NumberFormatter.java:
```diff
@@ -0,0 +1,220 @@
+  }
+
+  @Getter
+  private int decimalSize = -1;
```
Поля и методы не по канону расположены... я аж назад отмотал, думал где статик класс пропустил

theshadowco commented on engine/src/main/java/com/github/otymko/jos/runtime/NumberFormatter.java:
```diff
@@ -0,0 +1,220 @@
+import java.util.List;
+import java.util.Locale;
+
+public class NumberFormatter {
```
Возможно стоит использовать `@Data`

theshadowco commented on engine/src/main/java/com/github/otymko/jos/runtime/NumberFormatterPresentationParts.java:
```diff
@@ -0,0 +1,100 @@
+import java.util.ArrayList;
+import java.util.Collections;
+
+final class NumberFormatterPresentationParts {
```
Используй `@Value`

dmpas replied:
Класс не иммутабельный.

theshadowco commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,205 @@
+import java.util.Locale;
+import java.util.Map;
+
+public final class ValueFormatter {
```
`@UtilityClass`

theshadowco commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,205 @@
+      || param.charAt(i) == 't' && i + 1 < param.length() && param.charAt(i + 1) == 't') {
+        builder.setCharAt(i, 'a');
+        i++;
+        builder.delete(i, i + 1);
```
а здесь точно не упадет?

dmpas replied:
второй параметр у `delete` обозначе как exclusive, исключая. Так что удаляется только i-й элемент, чьё существование проверено условием `if`

EightM commented on engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java:
```diff
@@ -0,0 +1,216 @@
+        return Optional.of(paramList.get(name.toUpperCase()));
+      }
+    }
+    return Optional.empty();
```
```suggestion
    Arrays.stream(names)
      .map(name -> name.toUpperCase(Locale.ENGLISH))
      .map(paramList::get)
      .filter(Objects::nonNull)
      .findFirst();
```

dmpas replied:
а джава 8 так могёт?
`Locale.ENGLISH` сработает на рашшн-именах?

otymko replied:
Импорт главное не забыть)

EightM commented on engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java:
```diff
@@ -0,0 +1,216 @@
+    if (stringValue.isPresent()) {
+      return getLocale(stringValue.get());
+    }
+    return Locale.getDefault();
```
```suggestion
return get(names)
      .map(this::getLocale)
      .orElse(Locale.getDefault());
```

EightM commented on engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java:
```diff
@@ -0,0 +1,216 @@
+    parseParams();
+  }
+
+  public Optional<String> get(String[] names) {
```
Наверное лучше дать методу более понятное имя

dmpas replied:
`get` на мой взгляд вполне себе очевидное имя. Как в `Map`, `ArrayList`. Или ты к тому, что оно тут `Optional` ?

dmpas committed:
Update engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java

Co-authored-by: EightM <zchokobo@gmail.com>

EightM commented on engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java:
```diff
@@ -0,0 +1,216 @@
+    if (stringValue.isEmpty()) {
+      return Optional.empty();
+    }
+    return Optional.of(parseInt(stringValue.get()));
```
```suggestion
    return get(names)
        .map(stringValue -> parseInt(stringValue));
```

dmpas committed:
Update engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java

Co-authored-by: EightM <zchokobo@gmail.com>

EightM commented on engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java:
```diff
@@ -0,0 +1,216 @@
+    return Optional.of(parseInt(stringValue.get()));
+  }
+
+  public Optional<List<Integer>> getIntList(String[] names) {
```
Optional от коллекций не возвращают. Если что-то не так, можно просто возвратить пустую коллекцию

dmpas replied:
как тогда лучше поступить? Потому что есть принципиальная разница между незаданным параметром и параметром, устанавливающим пустой список. Не null же возвращать.
Или переделать на схему типа `keyPresent()` ?

otymko replied:
@dmpas Возвращай Collections.emptyList();

dmpas replied:
@otymko ну не. В Формате отсутствие `ЧГ` и `ЧГ=` - два совершенно разных формата.

EightM commented on engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java:
```diff
@@ -0,0 +1,216 @@
+    final var result = new ArrayList<Integer>();
+    final var stringValue = asStringValue.get();
+
+    for (final var groupElement : stringValue.split("\\D")) {
```
регулярку лучше вынести в статический Pattern, так как будет компилироваться при каждом вызове .split

EightM commented on engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java:
```diff
@@ -0,0 +1,216 @@
+      }
+    }
+
+    return Optional.of(result);
```
```suggestion
    get(names)
      .stream()
      .flatMap(stringValue -> Arrays.stream(stringValue.split("\\D")))
      .filter(Predicate.not(String::isBlank))
      .map(this::parseInt)
      .collect(Collectors.toList());
```

EightM commented on engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java:
```diff
@@ -0,0 +1,216 @@
+      return 0;
+    }
+
+    return Integer.parseInt(cleanValue.toString());
```
```suggestion
    String result = value.chars()
      .filter(ch -> Character.isDigit(ch) || ch == '-')
      .mapToObj(ch -> (char)ch)
      .map(String::valueOf)
      .collect(Collectors.joining());
      
      return Integer.parseInt(result)
```

dmpas committed:
Update engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java

Co-authored-by: EightM <zchokobo@gmail.com>

dmpas committed:
Update engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java

Co-authored-by: EightM <zchokobo@gmail.com>

EightM commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,205 @@
+
+public final class ValueFormatter {
+
+  private static final String[] BOOLEAN_FALSE = {"БЛ", "BF"};
```
Я бы заменил все массивы на List.of. Иммутабельность там, да и просто привычнее

EvilBeaver replied:
приватные массивы лучше листОф. Минимальная нагрузка на память, а приватность контролирует чужие ручонки

EightM commented on engine/src/main/java/com/github/otymko/jos/runtime/ValueFormatter.java:
```diff
@@ -0,0 +1,205 @@
+    return nf.format(value);
+  }
+
+  static {
```
статический блок не на своем месте

dmpas committed:
Update engine/src/main/java/com/github/otymko/jos/runtime/FormatParametersList.java

Co-authored-by: EightM <zchokobo@gmail.com>

dmpas committed:
Рефакторинг по замечаниям.

dmpas committed:
`List<>` вместо массивов.

dmpas committed:
Вынесен шаблон регулярки разбора списка.

dmpas committed:
Классы форматирования выделены в отдельный пакет.

dmpas committed:
Представление значений Булева локализовано.

dmpas committed:
Локализация значений теста.

dmpas committed:
Локализация значений теста.

dmpas committed:
JavaDoc публичных методов.

dmpas committed:
Merge remote-tracking branch 'oleg/develop' into feature/format

# Conflicts:
#	engine/src/main/java/com/github/otymko/jos/runtime/machine/MachineInstance.java
#	engine/src/test/java/com/github/otymko/jos/ExternalScriptTest.java

dmpas committed:
Переименовал локальную переменную.

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/format/FormatParametersList.java:
```diff
@@ -0,0 +1,76 @@
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+final class FormatParametersList {
```
@dmpas описания бы сюда, для чего это

otymko commented on engine/src/test/resources/tests/formatting.os:
```diff
@@ -0,0 +1,409 @@
+Процедура ТестДолжен_ПроверитьФорматированиеГруппировкаРазрядов() Экспорт
+
+	Число=123456789;
+	//юТест.ПроверитьРавенство("123,456,789", Формат(Число, "ЧГ; L=en_US"));
```
@dmpas давай на это заведем issue

otymko commented on engine/src/main/java/com/github/otymko/jos/runtime/format/FormatParametersListBuilder.java:
```diff
@@ -0,0 +1,154 @@
+import java.util.HashMap;
+import java.util.Map;
+
+final class FormatParametersListBuilder {
```
@dmpas описание бы сюда

dmpas committed:
Пропуск неверных параметров форматирования.

dmpas committed:
Формат: описание внутренних классов.

dmpas committed:
Формат: Рефакторинг разбора параметров.

dmpas committed:
Формат: Рефакторинг.
